### DISCUSSION TITLE
Question: Can I use a custom subdomain for preview deployments?

### DISCUSSION THREAD

**Discussion by dgokcin** [vercel/community#1056] [Help] [answered]
Hello

TL;DR: Can I point a subdomain to the **latest** preview deployment using vercel cli and/or aliases without the GitHub integration such that promoting a build to production does not overwrite anything?

I am using Cloudflare for managing my domain example.com. I have a vercel project set up without the Git Integration and I have two domains(example.com and qa.example.com) that I added to vercel. Both of the domains are shown as _**production**_ under project settings -> domains. 

When I build with `vercel build` and deploy with `vercel deploy  --prebuilt` on the _develop_ branch, and set an alias of the deployment to `qa.example.com`, this creates a preview deployment on Vercel UI, which is what I expect. However, if I build with `vercel build --prod`,`vercel deploy --prod --prebuilt` and set an alias to example.com on my _main_ branch, although example.com works as I expected, I see that qa.example.com looks the same as example.com. Is this expected and documented somewhere? If not how can I configure vercel to not do that?

Thanks

**Comment by amyegan** [comment, marked as answer]
Hi @dgokcin! It is expected that all custom `Production` [domains](https://vercel.com/docs/concepts/projects/domains#) take you to the latest production deployment. 

You can set up a [branch-specific subdomain](https://vercel.com/docs/concepts/projects/domains/assign-domain-to-a-git-branch#) as a staging domain. That wouldn't apply to all preview deployments, but would work for any deployed from a staging branch. Then `qa.example.com` would become a `Preview` domain and show the staging version of your site, while `example.com` would be your `Production` domain and show the production version.

<img width="734" alt="image" src="https://user-images.githubusercontent.com/8247105/202035635-b11f34c4-3677-4ea0-ae99-80201d34cc98.png">

You can learn more about how to set up a staging environment here: [vercel.com/guides/set-up-a-staging-environment-on-vercel](https://vercel.com/guides/set-up-a-staging-environment-on-vercel)

**Reply by dgokcin** [reply]
Thanks for the reply. One question though, will connecting my vercel project to git cause me additional costs in the pro plan(like 20USD per commiter)

**Reply by dgokcin** [reply]
@amyegan also if the answer is yes for the above question, would using `vercel deploy --prebuild` without the `--prod` option in both of my deployments(qa and prod) will cause a problem othet than not being able to promote a preview deployment from vercel ui?

**Reply by amyegan** [reply]
There is no additional charge for [connecting git repository to your Vercel Project](https://vercel.com/docs/concepts/git#). I use [Vercel for GitHub](https://vercel.com/docs/concepts/git/vercel-for-github#) a lot for my hobby projects and find it very convenient for most use cases. 

If you want to know more about the features that may cause additional charges, you can find more details on the [Limits](https://vercel.com/docs/concepts/limits/overview#), [Usage](https://vercel.com/docs/concepts/limits/usage#), and [Fair Use Policy](https://vercel.com/docs/concepts/limits/fair-use-policy#) pages.

---

I might be misunderstanding your second question. Please correct me if got it wrong!

All production builds can only be deployed to the production environment. And all preview builds can only be deployed to the preview environment. If you do a production build and try to deploy it without the `--prod` option then you will get an error like this:
```
Error: The "--prebuilt" option was used with the target environment "preview", but the prebuilt output found in ".vercel/output" was built with target environment "production". Please run `vercel --prebuilt --prod`.
Learn More: https://vercel.link/prebuilt-environment-mismatch
```

If you are comfortable with a more complex setup, the CLI can give you more flexibility for your preview deployment domain. You could use [`vercel alias`](https://vercel.com/docs/cli/alias#) to add a custom staging domain to your preview deployments. The production domain is automatically applied to the latest production deployment in any case.

There are some guides available to walk you through basic automation setup with Vercel CLI for [GitHub Actions](https://vercel.com/guides/how-can-i-use-github-actions-with-vercel), [GitLab Pipelines](https://vercel.com/guides/how-can-i-use-gitlab-pipelines-with-vercel), and [Bitbucket Pipelines](https://vercel.com/guides/how-can-i-use-bitbucket-pipelines-with-vercel) if you're interested in that approach.

**Reply by dgokcin** [reply]
Hello @amyegan thanks for the reply

So we are under the team plan and the domains are configured under the pro team , not under hobby personal account since we want the usage limitations in the personal account is not enough for us. What I read in [here](https://vercel.com/docs/concepts/git#using-teams) makes me think that If I am in the teams scope, I will get charged 20 USD per commiter if I enable the git integration even though I am building on GitHub actions deploying using the vercel CLI with the `--prebuild` argument. Is this correct or I can just connect my qa.example.com domain to the develop branch and deploy to the qa.example.com preview environment like the screenshots you have shared without any additional costs?

In this case, you can ignore my second question :). Would connecting qa.example.com to github develop branch wil result like the screenshot you have shared(qa.example.com -> preview) without any additional costs in the team plan after the following workflows execute(assuming that it already deploys correctly to the qa.example.com production domain)?

```yml
name: Deploy to Vercel QA
on:
  workflow_dispatch:
  repository_dispatch:
    types: deploy_to_qa
  push:
    branches:
      - develop

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

env:
  VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
  VERCEL_PROJECT_ID: ${{ secrets.VERCEL_MY_WEBSITE_PROJECT_ID }}
  VERCEL_TOKEN: ${{ secrets.VERCEL_GHA_TOKEN }}
  VERCEL_TEAM_ID: ${{ secrets.VERCEL_TEAM_ID }}
  NODE_VERSION: 16
  ENVIRONMENT: qa
  VERCEL_ENV: preview

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: ${{ github.event.client_payload.target_branch || github.ref }}
      - uses: actions/setup-node@v3
        with:
          node-version: ${{ env.NODE_VERSION }}
          # cache: yarn

      - name: Install dependencies
        run: yarn install --immutable

      - name: Pull Project Settings from Vercel
        run: yarn vercel pull --yes --environment=${{ env.VERCEL_ENV }} --token=${{ env.VERCEL_TOKEN }}

      - name: Pull Environment Variables from Vercel
        run: yarn vercel env pull --environment=${{ env.VERCEL_ENV }} --yes --token=${{ env.VERCEL_TOKEN }} .env.local

      - name: Move .env.${{ env.ENVIRONMENT }} to .env
        run: mv .env.${{env.ENVIRONMENT}} .env

      - name: Vercel Build
        run: yarn vercel build

      - uses: amondnet/vercel-action@v25 #deploy
        with:
          vercel-token: ${{ env.VERCEL_TOKEN }} # Required
          github-token: ${{ secrets.GITHUB_TOKEN }} #Optional
          vercel-args: --archive=tgz --prebuilt
          vercel-org-id: ${{ env.VERCEL_ORG_ID}}  #Required
          vercel-project-id: ${{ env.VERCEL_PROJECT_ID}} #Required
          scope: ${{ env.VERCEL_TEAM_ID }} #Optional
          alias-domains: |
            qa.example.com
```

```yml
name: Deploy to Vercel PROD
on:
  workflow_dispatch:
  repository_dispatch:
    types: deploy_to_prod
  push:
    branches:
      - main

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: false

env:
  VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
  VERCEL_PROJECT_ID: ${{ secrets.VERCEL_MY_WEBSITE_PROJECT_ID }}
  VERCEL_TOKEN: ${{ secrets.VERCEL_GHA_TOKEN }}
  VERCEL_TEAM_ID: ${{ secrets.VERCEL_TEAM_ID }}
  NODE_VERSION: 16
  ENVIRONMENT: prod
  VERCEL_ENV: production

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: ${{ github.event.client_payload.target_branch || github.ref }}
          fetch-depth: 0

      - name: Get Token
        id: get_workflow_token
        uses: peter-murray/workflow-application-token-action@v2
        with:
          application_id: ${{ secrets.APPLICATION_ID }}
          application_private_key: ${{ secrets.APPLICATION_PRIVATE_KEY }}

      - name: Bump version and push tag
        id: versioning
        uses: anothrNick/github-tag-action@v1
        env:
          GITHUB_TOKEN: ${{ steps.get_workflow_token.outputs.token }}
          WITH_V: true
          DEFAULT_BUMP: patch
          RELEASE_BRANCHES: main

      - uses: actions/setup-node@v3
        with:
          node-version: ${{ env.NODE_VERSION }}
          # cache: yarn

      - name: Install dependencies
        run: yarn install --immutable

      - name: Pull Project Settings from Vercel
        run: yarn vercel pull --yes --environment=${{ env.VERCEL_ENV }} --token=${{ env.VERCEL_TOKEN }}

      - name: Pull Environment Variables from Vercel
        run: yarn vercel env pull --environment=${{ env.VERCEL_ENV }} --yes --token=${{ env.VERCEL_TOKEN }} .env.local

      - name: Move .env.${{ env.ENVIRONMENT }} to .env
        run: mv .env.${{env.ENVIRONMENT}} .env

      - name: Vercel Build
        run: yarn vercel build

      - uses: amondnet/vercel-action@v25 #deploy
        with:
          vercel-token: ${{ env.VERCEL_TOKEN }} # Required
          github-token: ${{ secrets.GITHUB_TOKEN }} #Optional
          vercel-args: --archive=tgz --prebuilt --prod
          vercel-org-id: ${{ env.VERCEL_ORG_ID}}  #Required
          vercel-project-id: ${{ env.VERCEL_PROJECT_ID}} #Required
          scope: ${{ env.VERCEL_TEAM_ID }} #Optional
          alias-domains: |
            example.com
```

**Reply by amyegan** [reply]
Since [amondnet/vercel-action](https://github.com/amondnet/vercel-action) is not an official Vercel product, I'm not sure how it works. I'll explain what I can in a general sense, but the specifics may be a little different for that action.

With regard to costs, you are only charged the 20 USD for people you add to your team. 

The link you shared is talking about automatic deployments using [Vercel for Github](https://vercel.com/docs/concepts/git/vercel-for-github#). If the Vercel for Github integration attempts to deploy a commit from someone who is not a member of the team, the Deployment will be blocked and the commit author can send a request to join the team. If you accept the request, then that person would be added as a member of the team and you would be charged 20 USD for their seat. If you reject their request, then there is no additional charge but the deployment would remain blocked. 

It looks like the [`alias-domains`](https://github.com/amondnet/vercel-action#alias-domains) setting for [amondnet/vercel-action](https://github.com/amondnet/vercel-action) is meant to let you set the `qa.example.com` domain for any preview deployment. For more help with that action, you should reach out in their [discussions page](https://github.com/amondnet/vercel-action/discussions). They'll be able to give you more in-depth answers about how their features work.

**Reply by dgokcin** [reply]
@amyegan thank you very much for your reply and your patience. my question is not actually related with the amondnet's action since all it does is executing vercel cli and adding some metadata with the `-m` argument like below. I should have been more clear about what the action does. I am attaching the relevant output from GitHub logs.

## The vercel commands executed for the qa deployment

```
vercel --archive=tgz --prebuilt -t *** -m githubCommitSha=sha -m githubCommitAuthorName=dgokcin -m
githubCommitAuthorLogin=dgokcin -m githubDeployment=1 -m githubOrg=MyOrg -m githubRepo=my-website -m
githubCommitOrg=MyOrg -m githubCommitRepo=my-website -m githubCommitMessage="added an empty line to trigger a build
again" -m githubCommitRef=fix/vercel-preview-environment --scope ***
```
## The vercel commands used for setting alias for qa.example.com custom domain

```
vercel -t *** --scope *** alias https://my-website-some-random-characters-my-team.vercel.app/ qa.example.com
```

## My `vercel.json` that disables the Vercel Git integration 
```json
{
  "version": 2,
  "public": false,
  "github": {
    "enabled": false
  }
}
```

 So please correct me if I am wrong. The build is made with a github bot and the  deployment is made with the `--prebuild` option. qa.example.com custom domain is connected to `develop` branch from the domains menu(with [Vercel for GitHub](https://vercel.com/docs/concepts/git/vercel-for-github#) disabled with the above `vercel.json` will ensure that I am not charged per commiter in a repository and my builds and deployments with the `--prod` option will not overwrite preview builds, since the custom domain will show up as preview.

**Reply by amyegan** [reply]
It seems like your biggest concern is being charged per committer. A Pro plan is charged 20 USD for each member of the team. Team members on a Pro plan must be invited by a team owner. Otherwise, they will not become part of the team and will not have access to deploy or modify team projects. You can learn more about plan costs on the [pricing page](https://vercel.com/pricing). And you can learn more about [managing team roles](https://vercel.com/docs/teams-and-accounts/team-members-and-roles#managing-team-roles) in the [Teams and Accounts](https://vercel.com/docs/teams-and-accounts#) section of the docs. The guide on [what counts as an active user on Vercel](https://vercel.com/guides/what-is-an-active-user-on-vercel?query=bot#are-bots-active-users) is also useful.

Disabling GitHub in your `vercel.json` means that the Vercel for GitHub bot will never build the project. The setting has no effect on the number of people you add to your team.

It looks like your "Vercel Build" step is building the project via the CLI.

```
- name: Vercel Build
  run: yarn vercel build
```

And the [amondnet/vercel-action](https://github.com/amondnet/vercel-action) readme makes it look like they might be doing their own build step. Or maybe they just aren't expecting people to use the `--prebuilt` option. 😕 

If you're having trouble with your GitHub Workflow, you should take a look at the guide for [using GitHub Actions with Vercel](https://vercel.com/guides/how-can-i-use-github-actions-with-vercel). It has a simplified version of this to help you see the most important steps. I also think it's worth reaching out to the [amondnet/vercel-action folks](https://github.com/amondnet/vercel-action/discussions) for more guidance if you aren't sure how that action is deploying your project.

I know I'm throwing a lot of info and links at you, and this can all get very confusing. I hope this helped answer your questions. But please let me know what, if anything, is still not clear and I'll keep working to get you the information you need.

**Reply by dgokcin** [reply]
you have been super helpful, thanks!
